sdks/python/apache_beam/ml/inference/base.py

```diff
@@ -243,70 +244,273 @@ def share_model_across_processes(self) -> bool:
-    keys, unkeyed_batch = zip(*batch)
-    return zip(
-        keys, self._unkeyed.run_inference(unkeyed_batch, model, inference_args))
+    if not self._many_models:
+      keys, unkeyed_batch = zip(*batch)
+      return zip(
+          keys,
+          self._unkeyed.run_inference(unkeyed_batch, model, inference_args))
+
+    batch_by_key = {}
+    for key, example in batch:
+      if key not in batch_by_key:
+        batch_by_key[key] = []
+      batch_by_key[key].append(example)
+
+    predictions = []
+    for key, unkeyed_batches in batch_by_key.items():
+      id = self._key_to_id_map[key]
+      mh = self._id_to_mh_map[id]
+      keyed_model_tag = model.load(id)
+      keyed_model_shared_handle = multi_process_shared.MultiProcessShared(
+          mh.load_model, tag=keyed_model_tag)
+      keyed_model = keyed_model_shared_handle.acquire()
+      for inf in mh.run_inference(unkeyed_batches, keyed_model, inference_args):
+        predictions.append((key, inf))
+      keyed_model_shared_handle.release(keyed_model)
+
+    return predictions
 
   def get_num_bytes(self, batch: Sequence[Tuple[KeyT, ExampleT]]) -> int:
-    keys, unkeyed_batch = zip(*batch)
-    return len(pickle.dumps(keys)) + self._unkeyed.get_num_bytes(unkeyed_batch)
+    if not self._many_models:
+      keys, unkeyed_batch = zip(*batch)
+      return len(
+          pickle.dumps(keys)) + self._unkeyed.get_num_bytes(unkeyed_batch)
+    return len(pickle.dumps(batch))
 
   def get_metrics_namespace(self) -> str:
-    return self._unkeyed.get_metrics_namespace()
+    if not self._many_models:
+      return self._unkeyed.get_metrics_namespace()
+    return 'BeamML_KeyedModels'
 
   def get_resource_hints(self):
-    return self._unkeyed.get_resource_hints()
+    if not self._many_models:
+      return self._unkeyed.get_resource_hints()
+    return {}
 
   def batch_elements_kwargs(self):
-    return self._unkeyed.batch_elements_kwargs()
+    if not self._many_models:
+      return self._unkeyed.batch_elements_kwargs()
+    return {}
 
   def validate_inference_args(self, inference_args: Optional[Dict[str, Any]]):
-    return self._unkeyed.validate_inference_args(inference_args)
+    if not self._many_models:
+      return self._unkeyed.validate_inference_args(inference_args)
+    for mh in self._id_to_mh_map.values():
+      mh.validate_inference_args(inference_args)
 
   def update_model_path(self, model_path: Optional[str] = None):
-    return self._unkeyed.update_model_path(model_path=model_path)
+    if not self._many_models:
+      return self._unkeyed.update_model_path(model_path=model_path)
+    if model_path is not None:
+      raise RuntimeError(
+          'Model updates are currently not supported for ' +
+          'KeyedModelHandlers with multiple different per-key ' +
+          'ModelHandlers.')
 
   def get_preprocess_fns(self) -> Iterable[Callable[[Any], Any]]:
-    return self._unkeyed.get_preprocess_fns()
+    return []
```

Comment:
what's the rationale for this change?

Reply:
We don't allow these to have a non-zero length anyways (see the constructor). Them being non-empty was pointless before anyways, but I wanted to avoid having to special case the different types of unkeyed here.

Reply:
will this have an impact on chain calling into unkeyed preprocess funs for single or multi-model case?

Reply:
note that this is also a default implementation in the superclass.

Reply:
> will this have an impact on chain calling into unkeyed preprocess funs for single or multi-model case?

Yeah, though not significantly. It doesn't allow you to strictly chain unkeyed functions if you have a KeyedModelHandler, though you can pretty easily do things like: `with_preprocess_fn(lambda k, v: k, my_func(v))`. Note that this PR doesn't change that behavior

> note that this is also a default implementation in the superclass, so we can remove it entirely

Good call, done